### ISSUE TITLE
Second call of MQTTAsync_connect() fails always for IPv6 server URI

### ISSUE THREAD

**Issue by JuergenKosel** [eclipse-paho/paho.mqtt.c#179]
Hello,

if an IPv6 server URI without port number is used, than MQTTAsync_connect() could be used only once:
1. Call MQTTAsync_create()
   Output from gdb:
   `
   Breakpoint 1, MQTTAsync_create (handle=0x7179a02c, serverURI=0x717a6464 "tcp://[2002:3e92:37e2:0:250:56ff:fe28:1055]", 
     clientId=0x7179a030 "00:06:71:33:00:13", persistence_type=1, persistence_context=0x0)
     at paho.mqtt.c/src/MQTTAsync.c:506
   `
2. Then first call of  MQTTAsync_connect()
   Output from gdb:
   `
   Breakpoint 2, MQTTAsync_connect (handle=0x717d6e2c, options=0x717d7550) at paho.mqtt.c/src/MQTTAsync.c:2141
   2141    in paho.mqtt.c/src/MQTTAsync.c
   (gdb) next
   2142    in paho.mqtt.c/src/MQTTAsync.c
   (gdb) print *m
   $2 = {serverURI = 0x717d6f8c "[2002:3e92:37e2:0:250:56ff:fe28:1055]", ssl = 0, c = 0x717d7184, 
   cl = 0x5de01 <MqttBroker::MQTTAsync_connectionLost(void*, char*)>, 
   ma = 0x5dd99 <MqttBroker::MQTTAsync_messageArrived(void*, char*, int, MQTTAsync_message*)>, 
   dc = 0x5dda1 <MqttBroker::MQTTAsync_deliveryComplete(void*, int)>, context = 0x71799fe8, 
   `

With this call the connection to the broker could be established.
1. Restart the broker, so that the current connection will be lost and a new one could be established (if there is not this issue).
2. Another call of MQTTAsync_connect() should establish the connection:
   `
   Breakpoint 2, MQTTAsync_connect (handle=0x717d6e2c, options=0x717d7550) at paho.mqtt.c/src/MQTTAsync.c:2141
   2141    in paho.mqtt.c/src/MQTTAsync.c
   (gdb) next
   2142    in paho.mqtt.c/src/MQTTAsync.c
   (gdb) print *m
   $3 = {serverURI = 0x717d6f8c "[2002:3e92:37e2:0:250:56ff:fe28:1055", ssl = 0, c = 0x717d7184,
   `

This and all other attempts to reconnect fail, because the trailing ']' has been removed from the server URI.

**Comment by JuergenKosel** [comment]
Hello,

the problem is caused in function MQTTProtocol_addressPort(), file paho.mqtt.c/src/MQTTProtocolOut.c lines 69-70:
`if (buf[len - 1] == ']')
        buf[len - 1] = '\0';`

**Comment by icraggs** [comment]
Added a fix to the fixes branch.

**Comment by JuergenKosel** [comment]
Hello,

the problem is in function `MQTTProtocol_addressPort()` which modifies the `const char* uri`, with the temporary pointer `buf`. So the uri string can not be used another time!

**Comment by icraggs** [comment]
Sorry, I didn't add the actual fix to the commit.  It's there now, and merged into the develop branch too.

**Comment by icraggs** [comment]
I added a fix so that the uri isn't modified.  Thanks.
